ResNet Benchmark

Landanjs opened the pull request:
This PR adds the ResNet + ImageNet benchmark. This code should replicate our baseline numbers and includes the option to train using our three recipes. There is code for both local and streaming data. The code should be able to pull any ResNet specified by torchvision.

One major thing missing is FFCV support. I'm not sure if we want people to be using FFCV workflows and would add some complexity to the code base. It seems like decent effort to convert a standard dataset into an FFCV one. This means the mild  and possibly the medium and hot recipes will be dataloader bottlenecked. A potential alternative to FFCV would be NVIDIA's Dali, but that has yet to be scoped out.

I attempted to write tests that can be run using pytest within the `resnet` directory. Unfortunately, I wasn't able to make testing for streaming datasets since I'm not sure how to do this, is this necessary before merging? Otherwise, this works for the most part, but I had issues importing and using the `main()` function for the smoke test. The workaround was to add the current directory to `sys.path`, but I don't think this would be an issue if there was a `setup.py`. I think we need to set one up at some point

[Baseline WandB result](https://wandb.ai/mosaic-ml/landan-test/runs/35px60dy?workspace=user-landanjs): 76.47% accuracy in 3.06 hours

growlix reviewed (commented):
A few comments, mostly on the readme. We should probably add an FFCV version, but I'm not going to ask that of you.

Landanjs commented on resnet/README.md:
```diff
@@ -22,107 +22,172 @@
-   **Note:** This tutorial assumes that the dataset is installed to the `/tmp/ImageNet` path.
-
-1. The `Mild` and `Medium` recipes require converting the ImageNet dataset to FFCV format.  *This conversion step is only required to be performed once, once converted files can be stashed away for reuse with subsequent runs.*  The `Hot` recipe uses the original ImageNet data.
+# Prequisites
```
🤦

Landanjs commented on resnet/README.md:
```diff
@@ -22,107 +22,172 @@
+  * `composer`
+  * `streaming` - MosaicML's streaming dataset
+  * `wandb` - Weights and Biases for experiment tracking
+  * `omegaconf` - Configuration management
```
Added links to github repos. Also added a description for composer

Landanjs commented on resnet/data.py:
```diff
@@ -0,0 +1,219 @@
+IMAGENET_CHANNEL_MEAN = (0.485 * 255, 0.456 * 255, 0.406 * 255)
+IMAGENET_CHANNEL_STD = (0.229 * 255, 0.224 * 255, 0.225 * 255)
+
+class StreamingImageNet(Dataset, VisionDataset):
```
Yes, there is also a possibility that we remove data / model code from composer in the future and maintain these in benchmarks.

Landanjs commented on resnet/README.md:
```diff
@@ -22,107 +22,172 @@
+  * `wandb` - Weights and Biases for experiment tracking
+  * `omegaconf` - Configuration management
+
+# Test Dataloader
```
That's reasonable haha

Landanjs commented on resnet/README.md:
```diff
@@ -22,107 +22,172 @@
+composer --world_size 16 --node_rank 0 --master_addr 0.0.0.0 --master_port 7501 main.py yamls/resnet50.yaml
+
+# Node 1
+composer --world_size 16 --node_rank 1 --master_addr 0.0.0.0 --master_port 7501 main.py main.py yamls/resnet50.yaml
```
Nope, nice catch!

Landanjs commented on resnet/README.md:
```diff
@@ -22,107 +22,172 @@
+
+Now that you've installed dependencies and tested your dataset, let's start training!
+
+**Please remember**: edit the `path` and (if streaming) `local` paths in `resnet50.yaml` to point to your data.
```
👍

Landanjs commented on resnet/README.md:
````diff
@@ -22,107 +22,172 @@
+[epoch=0][batch=17/625]: metrics/train/Accuracy: 0.0010
+train          Epoch   0:    3%|▋                        | 17/625 [00:17<07:23,  1.37ba/s, loss/train/total=7.1292]
+```
+# Using Mosaic Recipes
````
I added a note at the bottom describing the FFCV sitch, let me know if it is enough!

growlix reviewed (approved):
LGTM!

dblalock commented on resnet/yamls/mcloud_run.yaml:
```diff
@@ -0,0 +1,95 @@
+  pip install -r requirements.txt
+  composer main.py /mnt/config/parameters.yaml
+
+# Configuration copied from baseline.yaml
+parameters:
```
Can this just not be here? Single specification on config would be a big readability, usability, and maintainability win.

Landanjs replied:
How are you thinking to reference `baseline.yaml` in this yaml? The other option in my mind is to do `composer main.py yamls/resnet50.yaml`. But this would mean that you would need to commit and push any changes to `resnet50.yaml` which I would find very annoying if I wanted to run something quickly.

I agree it would make readability and maintainability easier, but it seems like a big hit to usability.

dblalock commented on resnet/yamls/mcloud_run.yaml:
```diff
@@ -0,0 +1,95 @@
+run_name: resnet_example
+cluster: null  # Name of the cluster to use for this run
+gpu_type: null # Type of GPU to use
+gpu_num: null  # Number of GPUs to use specified as an integer
```
Can these be sane defaults? 
```suggestion
gpu_type: a100_40 # Type of GPU to use
gpu_num: 8  # Number of GPUs to use
```

Landanjs replied:
I was thinking about the [mcli examples](https://mcli.docs.mosaicml.com/getting_started/first_composer.html) where they don't specify gpu_type, but it's probably fine

dblalock commented on resnet/tests/test_trainer.py:
```diff
@@ -0,0 +1,45 @@
+
+from ..main import main
+from ..tests.utils import SynthClassificationDirectory
+
+
+
```
```suggestion


```

dblalock commented on resnet/tests/test_model.py:
```diff
@@ -0,0 +1,12 @@
+@pytest.mark.parametrize('loss_name', ['binary_cross_entropy', 'cross_entropy'])
+@pytest.mark.parametrize('num_classes', [10, 100])
+def test_model_builder(model_name, loss_name, num_classes):
+    model = build_composer_resnet(model_name, loss_name, num_classes)
+    assert isinstance(model, ComposerClassifier)
```
Can we also shove a (1, 3, 224, 244) randn tensor through it and check the output shape + dtype? Probably as a separate test

Landanjs replied:
Should it be a separate test? Feeding input and output shape should also test model creation, right?

Landanjs replied:
Added input-output tests

dblalock commented on resnet/data.py:
```diff
@@ -0,0 +1,219 @@
+IMAGENET_CHANNEL_STD = (0.229 * 255, 0.224 * 255, 0.225 * 255)
+
+class StreamingImageNet(Dataset, VisionDataset):
+    """ ImageNet streaming dataset based on PyTorch's VisionDataset.
+
```
```suggestion
    """ImageNet streaming dataset based on PyTorch's VisionDataset.

```

dblalock commented on resnet/data.py:
```diff
@@ -0,0 +1,219 @@
+        return image, target
+
+
+def build_streaming_imagenet_dataspec(
+    remote: str,
+    local: str,
+    batch_size: int,
+    is_train: bool = True,
+    drop_last: bool = True,
+    shuffle: bool = True,
+    resize_size: int = -1,
+    crop_size: int = 224,
+    **dataloader_kwargs,
+) -> DataSpec:
```
can we deduplicate `build_streaming_imagenet_dataspec` and `build_imagenet_dataspec`? You've thought about it more but it seems like we could just add a `streaming: bool = False` argument and just change the dataset type near the bottom?

Would save us and the user some code + make clear that using streaming datasets is barely different from what you're used to.

Landanjs replied:
Either works, I thought it was nice to have a clear distinction between using the two, but to your point, it might be good to blur the lines

Landanjs replied:
Done!

dblalock commented on resnet/main.py:
```diff
@@ -0,0 +1,238 @@
+from model import build_composer_resnet
+
+
+def build_logger(name, kwargs):
```
type annotations for function params? (also applies to other helper functions). Config is an omegaconf.DictConfig

Landanjs replied:
I think I added the right ones?

dblalock commented on resnet/main.py:
```diff
@@ -0,0 +1,238 @@
+
+def main(config):
+    if config.grad_accum == 'auto' and not torch.cuda.is_available():
+        raise ValueError('grad_accum="auto" requires training with a GPU, please specify grad_accum as an integer')
```
```suggestion
        raise ValueError('grad_accum="auto" requires training with a GPU; please specify grad_accum as an integer')
```

dblalock commented on resnet/main.py:
```diff
@@ -0,0 +1,238 @@
+
+    # Train dataset
+    print('Building train dataloader')
+    if config.train_dataset.is_streaming:
+        train_dataspec = build_streaming_imagenet_dataspec(
```
ya, this block makes me more convinced we should deduplicate the two dataspec builder functions--having separate functions lengthens both the mains training script and data.py

Landanjs replied:
Done!

dblalock commented on resnet/main.py:
```diff
@@ -0,0 +1,238 @@
+    # Callback for checkpointing
+    print('Built Speed, LR, and Memory monitoring callbacks\n')
+
+    # Recipes for training ResNet architectures on ImageNet in order of increasing training time and accuracy
+    # To learn about individual methods, check out "Methods Overview" in our documentation: https://docs.mosaicml.com/
+    print('Building algorithm recipes')
```
Don't we need to overwrite parts of the config based on the selected recipe? E.g., the recipe's train and eval crop sizes don't seem to be used.

Probably has to happen right when we load the config, at the bottom of the file.

I think [omegaconf accepts nested "." attributes in key names](https://omegaconf.readthedocs.io/en/2.2_branch/usage.html#omegaconf-update), so the overwriting should be pretty simple.

Landanjs replied:
I dropped the ball on this one. I was testing doing it at the bottom, but that would make it hard to test recipes in the `test_trainer` script. Any objections to adding it to the top of `main()`?

Landanjs replied:
I added the recipe update at the top of `main()` and added tests for the recipe names in `test_trainer`. I'm not sure if changes the parameters under the user is the right interface, but I think we should iterate on it instead of delay the PR

dblalock reviewed (commented):
This is awesome. Just a couple suggested changes to make stuff shorter.

dblalock commented on resnet/README.md:
```diff
@@ -22,107 +22,181 @@
 
-# Prequisites
+# Mosaic ResNet
+This folder contains starter code for training [torchvision ResNet architectures](https://pytorch.org/vision/stable/models.html) using our most efficient training recipes (see our [short blog](https://www.mosaicml.com/blog/mosaic-resnet) or [detailed blog](https://www.mosaicml.com/blog/mosaic-resnet-deep-dive) for details). These recipes were developed to hit baseline accuracy on [ImageNet](https ...
```
```suggestion
This folder contains starter code for training [torchvision ResNet architectures](https://pytorch.org/vision/stable/models.html) using our most efficient training recipes (see our [short blog post](https://www.mosaicml.com/blog/mosaic-resnet) or [detailed blog post](https://www.mosaicml.com/blog/mosaic-resnet-deep-dive) for details). These recipes were developed to hit baseline accuracy on [ImageNet](https://www.image-net.org/) 7x faster or to maximize ImageNet accuracy over long training durations. Although these recipes were developed for training ResNet on ImageNet, they could be used to train other image classification models on other datasets. Give it a try!
```

dblalock commented on resnet/README.md:
```diff
@@ -22,107 +22,181 @@
-
-First, choose a recipe you would like to work with: [`Mild`, `Medium`, `Hot`].  This will determine which configuration file, `recipe_yaml_path`, you will need to specify. 
+The specific files in this folder are:
+* `model.py` - A function to create a [ComposerModel](https://docs.mosaicml.com/en/v0.11.0/composer_model.html) from a torchvision ResNet model
+* `data.py` - A [MosaicML streaming dataset](https://docs.mosaicml.com/projects/streaming/en/latest/) for ImageNet and a PyTorch dataset for a local copy of ImageNet
+* `main.py` - A script that builds a [Composer](https://github.com/mosaicml/composer) Trainer based on a configuration specified by a yaml
```
```suggestion
* `model.py` - Creates a [ComposerModel](https://docs.mosaicml.com/en/v0.11.0/composer_model.html) from a torchvision ResNet model
* `data.py` - Provides a [MosaicML streaming dataset](https://docs.mosaicml.com/projects/streaming/en/latest/) for ImageNet and a PyTorch dataset for a local copy of ImageNet
* `main.py` - Trains a ResNet on ImagNet using the [Composer](https://github.com/mosaicml/composer) [Trainer](https://docs.mosaicml.com/en/stable/api_reference/generated/composer.Trainer.html#trainer). 
```

dblalock commented on resnet/README.md:
```diff
@@ -22,107 +22,181 @@
+
+This benchmark assumes that ImageNet is already stored on your local machine or stored in an S3 bucket after being processed into a streaming dataset. Instructions to download ImageNet can be found [here](https://www.image-net.org/download.php). Alternatively, you can train on other image classification datasets. The local dataset code assumes your data is in the [torchvision ImageFolder](https:/ ...
+
+The below command will test if your data is setup appropriately:
```
```suggestion
The below command will test if your data is set up appropriately:
```

dblalock commented on resnet/README.md:
```diff
@@ -22,107 +22,181 @@
+
+# Dataloader Testing
+
+This benchmark assumes that ImageNet is already stored on your local machine or stored in an S3 bucket after being processed into a streaming dataset. Instructions to download ImageNet can be found [here](https://www.image-net.org/download.php). Alternatively, you can train on other image classification datasets. The local dataset code assumes your data is in the [torchvision ImageFolder](https:/ ...
```
```suggestion
If you want to train on a dataset other than ImageNet, you'll need to either make it a [streaming dataset](https://github.com/mosaicml/streaming) using [this script](https://github.com/mosaicml/streaming/blob/86a9b95189e8b292a8c7880a1c49dc55d1895544/streaming/vision/convert/imagenet.py) or a local [torchvision ImageFolder](https://pytorch.org/vision/main/generated/torchvision.datasets.ImageFolder.html).
```

Landanjs replied:
Doesn't this apply to Imagenet as well? So would it be better to say "If you want to train on ImageNet or another dataset..."? I also think it would still be could to include ImageNet links

Landanjs replied:
I updated the description with my tweak

dblalock commented on resnet/README.md:
```diff
@@ -22,107 +22,181 @@
+# Using Mosaic Recipes
+
+As described in our [ResNet blog post](https://www.mosaicml.com/blog/mosaic-resnet), we cooked up three recipes to train ResNets faster and with higher accuracy:
+- **Mild** recipe is for shorter training runs
+- **Medium** recipe is for longer training runs
+- **Hot** recipe is for the very longest training runs that maximize accuracy
```
```suggestion
- The **Mild** recipe is for short training runs
- The **Medium** recipe is for longer training runs
- The **Hot** recipe is for the longest training runs, intended to maximize accuracy
```

dblalock commented on resnet/README.md:
````diff
@@ -22,107 +22,181 @@
-      wget -P /tmp https://raw.githubusercontent.com/mosaicml/composer/v0.7.1/scripts/ffcv/create_ffcv_datasets.py
-      ```
+To use a recipe, specify the name using the the `recipe_name` argument. Specifying a recipe will change several aspects of the training run such as:
+1. Changes the loss function to binary cross entropy instead of standard cross entropy since this has been shown to improve acurracy.
+2. Changes the train crop size to 176 (instead of 224) and the evaluation resize size to 232 (instead of 256). The smaller train crop size increases throughput and has been show to improve accuracy as well.
+3. Changes the number of training epochs to the optimal value for each training recipe. Feel free to change these in `resnet50.yaml` to better suite your model and/or dataset.
+4. Specifies unique sets of speedup methods for model training.
````
```suggestion
1. Changes the loss function to binary cross entropy instead of standard cross entropy to improve accuracy.
2. Changes the train crop size to 176 (instead of 224) and the evaluation size to 232 (instead of 256). The smaller train crop size increases throughput and accuracy.
3. Changes the number of training epochs to the optimal value for each training recipe. Feel free to change these in `resnet50.yaml` to better suite your model and/or dataset.
4. Specifies unique sets of [speedup methods](https://docs.mosaicml.com/en/stable/trainer/algorithms.html) for model training.
```

Landanjs replied:
This looks good! I think "evaluation resize size" is more accurate since the evaluation size is still 224, is it okay to keep it?

dblalock commented on resnet/README.md:
```diff
@@ -22,107 +22,181 @@
+4. Specifies unique sets of speedup methods for model training.
 
-   1. Convert the training and validation datasets.
+Here is an example command to run the mild recipe on a single-node:
```
```suggestion
Here is an example command to run the mild recipe on a single node:
```

dblalock commented on resnet/README.md:
```diff
@@ -22,107 +22,181 @@
 
-      **Note:** The helper script output the FFCV formatted dataset files to `/tmp/imagenet_train.ffcv` and `/tmp/imagenet_val.ffcv` 
-      for the training and validation data, respectively.
+The throughput of ResNet50 and smaller models are dataloader bottlenecked when training with our recipes on 8x NVIDIA A100s. This means the model's throughput is limited to the speed the data can be loaded. One way to alleviate this bottleneck is to use the [FFCV dataloader format](https://github.com/libffcv/ffcv). This [tutorial](https://docs.mosaicml.com/en/v0.11.0/examples/ffcv_dataloaders.htm ...
```
```suggestion
The ResNet50 and smaller models are dataloader-bottlenecked when training with our recipes on 8x NVIDIA A100s. This means the model's throughput is limited to the speed the data can be loaded. One way to alleviate this bottleneck is to use the [FFCV dataloader format](https://github.com/libffcv/ffcv). This [tutorial](https://docs.mosaicml.com/en/v0.11.0/examples/ffcv_dataloaders.html) walks you through creating your FFCV dataloader. We also have example code for an ImageNet FFCV dataloader [here](https://github.com/mosaicml/composer/blob/a0f441537008a1ef2678f1474f3cd5519deb80fa/composer/datasets/imagenet.py#L179).
```

dblalock commented on resnet/README.md:
```diff
@@ -22,107 +22,181 @@
+# Saving and Loading checkpoints
 
-   Replace `num_gpus`, `recipe_yaml_path` and `scale_schedule_ratio` with the total number of GPU's, the recipe configuration, and the scale schedule ratio we determined in the previous section for the desired run, respectively.
+At the bottom of `yamls/resnet50.yaml`, we provide arguments for saving and loading model weights. Please specify the `save_folder` or `load_path`arugments if you need to save or load checkpoints!
```
```suggestion
At the bottom of `yamls/resnet50.yaml`, we provide arguments for saving and loading model weights. Please specify the `save_folder` or `load_path` arguments if you need to save or load checkpoints!
```

Also can we just set some sane defaults here? Like `./checkpoints` ?

Landanjs replied:
`null` save path means don't checkpoint, I don't think we would want to checkpoint by default, right?

dblalock commented on resnet/README.md:
````diff
@@ -22,107 +22,181 @@
-   composer -n 8 train.py -f recipes/resnet50_mild.yaml --scale_schedule_ratio 0.32
-   ```
+# On memory constraints
+In previous blogs ([1](https://www.mosaicml.com/blog/farewell-oom), [2](https://www.mosaicml.com/blog/billion-parameter-gpt-training-made-easy))
````
```suggestion
In previous blog posts ([1](https://www.mosaicml.com/blog/farewell-oom), [2](https://www.mosaicml.com/blog/billion-parameter-gpt-training-made-easy))
```

dblalock commented on resnet/README.md:
````diff
@@ -22,107 +22,181 @@
-   ```
+# On memory constraints
+In previous blogs ([1](https://www.mosaicml.com/blog/farewell-oom), [2](https://www.mosaicml.com/blog/billion-parameter-gpt-training-made-easy))
+we demonstrated Auto Grad Accum, which allows Composer to determine `grad_accum` on its own. This means the same configuration can be run on different hardware or on a fewer number of devices without having to manually adjust the gradient accumulation! We have done extensive testing on this feature, but if there are any issues you can manually set `grad_accum` to your desired value.
````
I doubt most users know what `grad_accum` is. We need a sentence explaining it if we're going to talk about (which maybe we don't need to?)

Landanjs replied:
I just took this from `llm`, I'm fine with deleting

Landanjs replied:
I deleted it

Landanjs commented:
@dblalock I believe I addressed most of your comments, but let me know if I missed anything! I wasn't sure about a few of the comments, so I left those open. Let me know what you think.

dblalock reviewed (approved):
LGTM other than the one redundant config thing, but we can change that in a separate PR--especially since every benchmark will hit this problem.
